lib_i2s/src/i2s_frame_slave_impl.h

```diff
@@ -74,85 +85,210 @@ static void i2s_frame_slave0(client i2s_frame_callback_if i2s_i,
 
-
-            if (num_out && (restart == I2S_NO_RESTART)){
+            if (num_out && (restart == I2S_NO_RESTART))
+            {
                 i2s_i.send(num_out << 1, out_samps);
 
-                //Output i2s evens (0,2,4..)
+                // Output i2s evens (0,2,4..)
+                if (num_data_bits == 32)
+                {
 #pragma loop unroll
-                for (size_t i=0, idx=0; i<num_out; i++, idx+=2){
-                    p_dout[i] <: bitrev(out_samps[idx]);
+                    for (size_t i = 0, idx = 0; i < num_out; i++, idx += 2)
+                    {
+                        p_dout[i] <: bitrev(out_samps[idx]);
+                    }
+                }
+                else
+                {
+#pragma loop unroll
+                    for (size_t i = 0, idx = 0; i < num_out; i++, idx += 2)
+                    {
+                        partout(p_dout[i], num_data_bits, bitrev(out_samps[idx] << data_bit_offset));
+                    }
                 }
             }
                 
-            //Read lrclk value
-            p_lrclk :> lrval;
+            // Read lrclk value
+            if (num_data_bits == 32)
+            {
+                p_lrclk :> lrval;
+            }
+            else
+            {
+                asm volatile("in %0, res[%1]"
+                            :"=r"(lrval)
+                            :"r"(p_lrclk));
+                set_port_shift_count(p_lrclk, num_data_bits);
+            }
 
-            //Input i2s evens (0,2,4..)
+            // Input i2s evens (0,2,4..)
+            if (num_data_bits == 32)
+            {
 #pragma loop unroll
-            for (size_t i=0, idx=0; i<num_in; i++, idx+=2){
-                int32_t data;
-                asm volatile("in %0, res[%1]":"=r"(data):"r"(p_din[i]):"memory");
-                in_samps[idx] = bitrev(data);
+                for (size_t i = 0, idx = 0; i < num_in; i++, idx += 2)
+                {
+                    int32_t data;
+                    asm volatile("in %0, res[%1]"
+                                :"=r"(data)
+                                :"r"(p_din[i]));
+                    in_samps[idx] = bitrev(data);
+                }
+            }
+            else
+            {
+#pragma loop unroll
+                for (size_t i=0, idx=0; i<num_in; i++, idx+=2)
+                {
+                    int32_t data;
+                    asm volatile("in %0, res[%1]"
+                                :"=r"(data)
+                                :"r"(p_din[i]));
+                    set_port_shift_count(p_din[i], num_data_bits);
+                    in_samps[idx] = bitrev(data) & data_bit_mask;
+                }
             }
 
-            syncerror += (lrval != expected_low);
-
-            //Read lrclk value
-            p_lrclk :> lrval;
+                
+            syncerror += ((lrval & !data_bit_mask) != expected_low);
+            
+            
+            if (num_data_bits == 32)
```

Comment:
Wondering why we need this special case? The else block should handle things properly?

Reply:
So the else block does handle things properly, but not the most efficiently. Between the added calls to `setpsc`, added bitmasking, etc. there's a bit of a hit on overhead in the general case which we can shave off by using previous behaviour in the case that `num_data_bits == 32` - which will be the majority of applications anyway